File: goVersion/array/arrayOp.go
package main

import (
	"fmt"
	"strings"
)

// NodeA - узел массива
type NodeA struct {
	value string
}

// Array - структура динамического массива
type Array struct {
	data     []NodeA
	size     int
	capacity int
}

// CreateArray - создание нового массива
func CreateArray() *Array {
	return &Array{
		capacity: 1,
		data:     make([]NodeA, 1),
		size:     0,
	}
}

// CreateNodeAr - создание нового узла
func CreateNodeAr(value string) *NodeA {
	return &NodeA{value: value}
}

// PushBack - добавление элемента в конец массива
func (arr *Array) PushBack(value string) {
	if arr.size >= arr.capacity {
		arr.capacity *= 2
		newData := make([]NodeA, arr.capacity)
		copy(newData, arr.data[:arr.size])
		arr.data = newData
	}

	if arr.size < len(arr.data) {
		arr.data[arr.size] = NodeA{value: value}
	} else {
		arr.data = append(arr.data, NodeA{value: value})
	}
	arr.size++
}

// AddInx - добавление элемента по индексу
func (arr *Array) AddInx(value string, inx int) error {
	if inx < 0 || inx > arr.size {
		return fmt.Errorf("выход за диапазон")
	}

	if arr.size >= arr.capacity {
		arr.capacity *= 2
		newData := make([]NodeA, arr.capacity)
		copy(newData, arr.data[:arr.size])
		arr.data = newData
	}

	// Сдвигаем элементы вправо
	arr.data = append(arr.data, NodeA{})
	if inx < arr.size {
		copy(arr.data[inx+1:], arr.data[inx:arr.size])
	}
	arr.data[inx] = NodeA{value: value}
	arr.size++

	return nil
}

// PrintArray - вывод массива
func (arr *Array) PrintArray() {
	if arr.IsEmpty() {
		fmt.Println("Массив пустой")
		return
	}

	fmt.Print("Вывод массива: ")
	for i := 0; i < arr.size; i++ {
		fmt.Printf("%s ", arr.data[i].value)
	}
	fmt.Println()
}

// GetInx - получение элемента по индексу
func (arr *Array) GetInx(inx int) (string, error) {
	if inx < 0 || inx >= arr.size {
		return "", fmt.Errorf("выход за диапазон")
	}
	return arr.data[inx].value, nil
}

// ChangeInx - замена элемента по индексу
func (arr *Array) ChangeInx(newValue string, inx int) error {
	if inx < 0 || inx >= arr.size {
		return fmt.Errorf("выход за диапазон")
	}
	arr.data[inx].value = newValue
	return nil
}

// SizeArr - получение размера массива
func (arr *Array) SizeArr() int {
	return arr.size
}

// RemoveElArr - удаление элемента по индексу
func (arr *Array) RemoveElArr(inx int) (string, error) {
	if inx < 0 || inx >= arr.size {
		return "", fmt.Errorf("выход за диапазон")
	}

	removedValue := arr.data[inx].value

	// Сдвигаем элементы влево
	copy(arr.data[inx:], arr.data[inx+1:arr.size])
	arr.size--

	return removedValue, nil
}

// PrintAllSubarrays - вывод всех различных подмассивов
func (arr *Array) PrintAllSubarrays() {
	n := arr.size

	fmt.Println("Все различные подмассивы:")
	
	// Пустой подмассив
	fmt.Println("{}")

	// Генерируем все возможные подмассивы
	for start := 0; start < n; start++ {
		for end := start; end < n; end++ {
			var subarray []string
			for i := start; i <= end; i++ {
				subarray = append(subarray, arr.data[i].value)
			}
			fmt.Printf("{%s}\n", strings.Join(subarray, ", "))
		}
	}
}

// IsEmpty - проверка на пустоту
func (arr *Array) IsEmpty() bool {
	return arr.size == 0
}

File: array/arrayOp.cpp
#include<iostream>
#include"arrayOp.h"
#include <string>
#include <vector>
using namespace std;

bool empty(Array* arr){
    return arr->size == 0;    
}

//создание массива
void createArray(Array* arr){
    arr->capacity = 1;
    arr->data = new NodeA[arr->capacity];
    arr->size = 0;
}

NodeA* createNodeAr(const string& value){
    NodeA* newNode = new NodeA;
    newNode->value = value; //задаём значение
    return newNode; //возвращаем указатель
}

void pushBack(Array* arr, const string& value){
    if (arr->size >= arr->capacity) {//проверяем нужно ли увеличить емкость
        arr->capacity *= 2;
        NodeA* newData = new NodeA[arr->capacity];
        
        //копируем существующие элементы
        for (int i = 0; i < arr->size; i++) {
            newData[i] = arr->data[i];
        }
        
        delete[] arr->data;
        arr->data = newData;
    }
    
    NodeA newNode;
    newNode.value = value;
    arr->data[arr->size] = newNode;
    arr->size++;
}

void addInx(Array* arr, const string& value, int inx){
    if (inx < 0 || inx > arr->size){
        cout << "Выход за диапазон" << endl;
        return;
    }
    
    if (arr->size >= arr->capacity) {
        arr->capacity *= 2;
        NodeA* newData = new NodeA[arr->capacity];
        
        for (int i = 0; i < arr->size; i++) {
            newData[i] = arr->data[i];
        }
        
        delete[] arr->data;
        arr->data = newData;
    }
    
    //сдвигаем элементы вправо, начиная с конца
    for (int i = arr->size; i > inx; i--) {
        arr->data[i] = arr->data[i-1];
    }
    
    NodeA newNode;
    newNode.value = value;
    arr->data[inx] = newNode;
    arr->size++;
}

void printArray(Array* arr){
    if (empty(arr)){
        cout << "Массив пустой" << endl;
        return;
    }

    cout << "Вывод массива: ";
    for (int i = 0; i < arr->size; i++) {
        cout << arr->data[i].value << " ";
    }
    cout << endl;
}

//получение эл-та по индексу
string getInx(Array* arr, int inx){
    if (inx < 0 || inx >= arr->size){
        cout << "Выход за диапазон" << endl;
        return "";
    }
    return arr->data[inx].value;
}

//замена эл-та по индексу
void changeInx(Array* arr, const string& newValue, int inx){
    if (inx < 0 || inx >= arr->size){
        cout << "Выход за диапазон" << endl;
        return;
    }
    arr->data[inx].value = newValue;
}

int sizeArr(Array* arr){
    return arr->size;
}

string removeElArr(Array* arr, int inx){
    if (inx < 0 || inx >= arr->size){
        cout << "Выход за диапазон" << endl;
        return "";
    }

    //сохраняем значение удаляемого элемента
    string removedValue = arr->data[inx].value;
    
    //сдвигаем элементы влево, начиная с удаляемого индекса
    for (int i = inx; i < arr->size - 1; i++) {
        arr->data[i] = arr->data[i+1];
    }
    
    arr->size--;
    return removedValue;
}

void printAllSubarrays(Array* arr) {
    int n = arr->size;
    int total = 1 << n; //двоичная форма 2^n
    
    cout << "Все различные подмножества:" << endl;
    
    for (int mask = 0; mask < total; mask++) {
        cout << "{";
        bool first = true;
        for (int i = 0; i < n; i++) {
            if (mask & (1 << i)) { // двоичная форма 000, 001 и т.д
                if (!first) cout << ", ";
                cout << arr->data[i].value;
                first = false;
            }
        }
        cout << "}" << endl;
    }
}


//0 - 000 - нет - {}
//1 - 001 - data[2] - {z}
//2 - 010 - data[1] - {y}
//3 - 011 - data[1], data[2] - {y, z}
//4 - 100 - data[0] - {x}
//5 - 101 - data[0], data[2] - {x, z}
//6 - 110 - data[0], data[1] - {x, y}
//7 - 111 - все - {x, y, z}